src/common/snippets/src/lowered/pass/insert_buffers.cpp

```diff
@@ -163,17 +163,33 @@ void InsertBuffers::insertion(LinearIR& linear_ir, const LinearIR::LoopManagerPt
                     const auto buffer_consumers_inputs = buffer_out->get_consumers();
                     linear_ir.replace_input(buffer_consumers_inputs, output_connector);
                     potential_consumers.insert(buffer_consumers_inputs.begin(), buffer_consumers_inputs.end());
-                    linear_ir.erase(std::find(linear_ir.begin(), linear_ir.end(), buffer));
+                    const auto buffer_pos = std::find(linear_ir.cbegin(), linear_ir.cend(), buffer);
+                    OPENVINO_ASSERT(buffer_pos != linear_ir.cend(), "Buffer has not been found in Linear IR");
+                    linear_ir.erase(buffer_pos);
                 }
             }
 
+            // potential_consumers is unsorted by linear IR set.
+            // We have to find first expr in Linear IR from the set to insert Buffer before *all* consumers
+            OPENVINO_ASSERT(!potential_consumers.empty(), "Buffer should have one consumer at least");
+            auto consumer_expr = potential_consumers.begin()->get_expr();
+            if (potential_consumers.size() > 1) {
+                std::set<ExpressionPtr> consumers;
+                for (const auto& port : potential_consumers)
+                    consumers.insert(port.get_expr());
+                const auto it = std::find_if(linear_ir.cbegin(), linear_ir.cend(),
```

Comment:
It looks really expensive to go through the entire IR just to find the first consumer. Especially considering the fact that it would be repeated for every loop exit (BTW what about loop inputs? don't we have a similar problem there?).
Can we at least start from the source expression itself to avoid passing through all the expr before the source?